hibernate-core/src/main/java/org/hibernate/query/Query.java

```diff
@@ -37,63 +37,43 @@
 import jakarta.persistence.TypedQuery;
 
 /**
- * Represents an HQL/JPQL query or a compiled Criteria query.  Also acts as the Hibernate
- * extension to the JPA Query/TypedQuery contract
+ * Represents an HQL or Criteria query.
+ *
+ * Hibernate extension of {@link TypedQuery}, adapting to the
+ * {@link org.hibernate.query.CommonQueryContract} hierarchy
  *
  * @author Gavin King
  * @author Steve Ebersole
  *
  * @param <R> The query result type, for typed queries, or {@code Object} for untyped queries
  */
 @Incubating
-public interface Query<R> extends TypedQuery<R>, CommonQueryContract {
-	/**
-	 * Get the QueryProducer this Query originates from.  Generally speaking,
-	 * this is the Session/StatelessSession that was used to create the Query
-	 * instance.
-	 *
-	 * @return The producer of this query
-	 */
-	SharedSessionContract getSession();
-
-	/**
-	 * Get the query string.  Note that this may be {@code null} or some other
-	 * less-than-useful return because the source of the query might not be a
-	 * String (e.g., a Criteria query).
-	 *
-	 * @return the query string.
-	 */
-	String getQueryString();
+public interface Query<R> extends SelectionQuery, JpaQuery, TypedQuery<R> {
```

Comment:
I don't get why `SelectionQuery` isn't `SelectionQuery<R>` instead?

I'm not sure raw types (e.g. in `getResultList()`) will bring any value to users. They'll be better off with typed results, and *when* we need untyped queries in the API, we can use `SelectionQuery<Object>` or `SelectionQuery<?>` (that's what wildcards are for). Unless there's an actual reason to use raw types that I didn't catch... ?

Reply:
```
package jakarta.persistence;

public interface Query {
   ...
}
```

Reply:
Sure, but AFAICS, all type that implement `SelectionQuery` _extend_ JPA's Query type. JPA's Query type doesn't extend `SelectionQuery`. So you're free to "do better" than JPA here. For example by adding generics: `SelectionQuery<R>`. Just like you did in `org.hibernate.query.Query<R>`, or like JPA did in `TypedQuery`, really.

Or did using generics lead to problems related to contravariance of R, e.g. a method of `SelectionQuery` accepting a parameter of type `R`? I'd be surprised, but if such a problem arises, there are other solutions than just sticking to raw types.